### PR TITLE
refactor: [Checkout] change payment and webhooks API contract

### PR DESCRIPTION
**Pull request by swangi-kumari**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Resolves [4486](https://github.com/juspay/hyperswitch-cloud/issues/4486)

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
Create a Card Non 3ds and 3ds payment and do refund for the same. Also configure webhooks and check for all the events.

On Checkout dashboard you can configure Webhooks from` Developer -> Webhooks -> New Webhook`
Under events enable `Disputes` and `Gateway` Events

Payment Create for card no 3ds
```
{
    "amount": 100000,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 100000,
    "customer_id": "StripeCustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://duck.com",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
      "card_number": "5385308360135181",
      "card_exp_month": "10",
      "card_exp_year": "25",
      "card_holder_name": "Joseph Doe",
      "card_cvc": "123" 
    }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "PiX"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "PiX"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}
```
Paymnet Response for Card No 3ds
```
{
    "payment_id": "pay_EhIzTPz73pEcipFaVHvH",
    "merchant_id": "merchant_1709891938",
    "status": "processing",
    "amount": 100000,
    "net_amount": 100000,
    "amount_capturable": 0,
    "amount_received": null,
    "connector": "checkout",
    "client_secret": "pay_EhIzTPz73pEcipFaVHvH_secret_nRSE9HFvr1xsJ1MFI9I9",
    "created": "2024-03-08T11:36:02.581Z",
    "currency": "USD",
    "customer_id": "StripeCustomer",
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "5181",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "538530",
            "card_extended_bin": "53853083",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "Joseph Doe"
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "PiX",
            "last_name": null
        },
        "phone": {
            "number": null,
            "country_code": null
        },
        "email": null
    },
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "PiX",
            "last_name": null
        },
        "phone": {
            "number": null,
            "country_code": null
        },
        "email": null
    },
    "order_details": null,
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://duck.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": null,
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "StripeCustomer",
        "created_at": 1709897762,
        "expires": 1709901362,
        "secret": "epk_7641d4ed9d3b472db8e20f865730b54a"
    },
    "manual_retry_allowed": false,
    "connector_transaction_id": "pay_tnmq7gnqtmaehlt2u5xaxn6wey",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pay_EhIzTPz73pEcipFaVHvH_1",
    "payment_link": null,
    "profile_id": "pro_be1gUpysocx6W37SPG4t",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_Pn5Xfkl0OhRYI8v8hwkf",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "request_external_3ds_authentication": null,
    "expires_on": "2024-03-08T11:51:02.581Z",
    "fingerprint": null
}
```
Do Payment Retrieve for the same Payment ID , status should be "succeeded"

Paymnet Request For Card 3ds
```
{
  "amount": 10000,
  "currency": "USD",
  "confirm": true,
  "capture_method": "automatic",
  "capture_on": "2022-09-10T10:11:12Z",
  "customer_id": "StripeCustomer",
  "email": "abcdef123@gmail.com",
  "name": "John Doe",
  "phone": "999999999",
  "phone_country_code": "+65",
  "description": "Its my first payment request",
  "authentication_type": "three_ds",
  "return_url": "https://google.com",
  "setup_future_usage": "off_session",
  "billing": {
    "address": {
      "line1": "1467",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "94122",
      "country": "US"
    }
  },
  "browser_info": {
    "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
    "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
    "language": "nl-NL",
    "color_depth": 24,
    "screen_height": 723,
    "screen_width": 1536,
    "time_zone": 0,
    "java_enabled": true,
    "java_script_enabled": true,
    "ip_address": "127.0.0.1"
  },
  "shipping": {
    "address": {
      "line1": "1467",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "94122",
      "country": "US",
      "first_name": "John",
      "last_name": "Doe"
    }
  },
  "statement_descriptor_name": "joseph",
  "statement_descriptor_suffix": "JS",
  "metadata": {
    "udf1": "value1",
    "new_customer": "true",
    "login_date": "2019-09-10T10:11:12Z"
  },
  "payment_method": "card",
  "payment_method_data": {
    "card": {
      "card_number": "5385308360135181",
      "card_exp_month": "10",
      "card_exp_year": "25",
      "card_holder_name": "Joseph Doe",
      "card_cvc": "123"
    }
  }
}
```
Payment Response for Card 3ds
```
{
    "payment_id": "pay_z8U0xwgUnYD2BTDKHLaz",
    "merchant_id": "merchant_1709891938",
    "status": "requires_customer_action",
    "amount": 10000,
    "net_amount": 10000,
    "amount_capturable": 10000,
    "amount_received": null,
    "connector": "checkout",
    "client_secret": "pay_z8U0xwgUnYD2BTDKHLaz_secret_c1BdGVuoXDPiHC1juzfO",
    "created": "2024-03-08T11:41:58.987Z",
    "currency": "USD",
    "customer_id": "StripeCustomer",
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": "off_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "card",
    "payment_method_data": {
        "card": {
            "last4": "5181",
            "card_type": null,
            "card_network": null,
            "card_issuer": null,
            "card_issuing_country": null,
            "card_isin": "538530",
            "card_extended_bin": "53853083",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "Joseph Doe"
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "John",
            "last_name": "Doe"
        },
        "phone": {
            "number": null,
            "country_code": null
        },
        "email": null
    },
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": null,
            "last_name": null
        },
        "phone": {
            "number": null,
            "country_code": null
        },
        "email": null
    },
    "order_details": null,
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": {
        "type": "redirect_to_url",
        "redirect_to_url": "http://localhost:8080/payments/redirect/pay_z8U0xwgUnYD2BTDKHLaz/merchant_1709891938/pay_z8U0xwgUnYD2BTDKHLaz_1"
    },
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": null,
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "StripeCustomer",
        "created_at": 1709898118,
        "expires": 1709901718,
        "secret": "epk_b5fbfc761c114252bfbe35f0d7072cf6"
    },
    "manual_retry_allowed": null,
    "connector_transaction_id": "pay_tpmddusbl3hulmsrkzyzlbb3xa",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pay_z8U0xwgUnYD2BTDKHLaz_1",
    "payment_link": null,
    "profile_id": "pro_be1gUpysocx6W37SPG4t",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_Pn5Xfkl0OhRYI8v8hwkf",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "request_external_3ds_authentication": null,
    "expires_on": "2024-03-08T11:56:58.987Z",
    "fingerprint": null
}
```
Do Payment Retrieve for the same Payment ID, status should be "succeeded"

Do Refund for same Payment ID
```
{
    "payment_id": "{{payment_id}}",
    "amount": 100,
    "reason": "Customer returned product",
    "refund_type": "instant",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    }
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by ArjunKarthik — DISMISSED** [review]

File: crates/router/src/connector/checkout.rs [hunk 1]
```diff
@@ -1244,23 +1244,32 @@ impl api::IncomingWebhook for Checkout {
...
                         .action_id
                         .ok_or(errors::ConnectorError::WebhookReferenceIdNotFound)?,
                 ),
+            };
+
+            return Ok(api_models::webhooks::ObjectReferenceId::RefundId(
+                refund_reference,
             ));
         }
         Ok(api_models::webhooks::ObjectReferenceId::PaymentId(
```
Comment (SamraatBansal) [inline crates/router/src/connector/checkout.rs:1275]: Try to handle with Hyperswitch reference attempt_id first, if it is not provided use checkout's connector transaction id

File: crates/router/src/connector/checkout/transformers.rs [hunk 2]
```diff
@@ -446,9 +451,16 @@ impl TryFrom<CheckoutWebhookEventType> for CheckoutPaymentStatus {
...
+            | CheckoutWebhookEventType::AuthenticationApproved
+            | CheckoutWebhookEventType::AuthenticationAttempted => Ok(Self::Pending),
+            CheckoutWebhookEventType::AuthenticationExpired
+            | CheckoutWebhookEventType::AuthenticationFailed
+            | CheckoutWebhookEventType::PaymentAuthenticationFailed
+            | CheckoutWebhookEventType::PaymentCaptureDeclined => Ok(Self::Declined),
             CheckoutWebhookEventType::PaymentRefunded
             | CheckoutWebhookEventType::PaymentRefundDeclined
+            | CheckoutWebhookEventType::PaymentCanceled
+            | CheckoutWebhookEventType::PaymentVoided
```
Comment (SamraatBansal) [inline crates/router/src/connector/checkout/transformers.rs:463]: Handle these status in Psync then map it here

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

**Review by SamraatBansal — APPROVED** [review]

File: crates/router/src/connector/checkout/transformers.rs [hunk 3]
```diff
@@ -450,6 +453,16 @@ pub enum CheckoutPaymentStatus {
     CardVerified,
     Declined,
     Captured,
+    #[serde(rename = "Retry Scheduled")]
+    RetryScheduled,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/checkout/transformers.rs:457]: Please raise this with connector and get clarity on when do we get this status?

File: crates/router/src/connector/checkout/transformers.rs [hunk 4]
```diff
@@ -494,10 +514,16 @@ impl ForeignFrom<(CheckoutPaymentStatus, Option<enums::CaptureMethod>)> for enum
...
+            | CheckoutPaymentStatus::PartiallyRefunded
+            | CheckoutPaymentStatus::Refunded => Self::Charged,
+            CheckoutPaymentStatus::PartiallyCaptured => Self::PartialCharged,
+            CheckoutPaymentStatus::Declined | CheckoutPaymentStatus::Expired => Self::Failure,
             CheckoutPaymentStatus::Pending => Self::AuthenticationPending,
-            CheckoutPaymentStatus::CardVerified => Self::Pending,
+            CheckoutPaymentStatus::CardVerified | CheckoutPaymentStatus::RetryScheduled => {
+                Self::Pending
+            }
+            CheckoutPaymentStatus::Voided | CheckoutPaymentStatus::Canceled => Self::Voided,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/checkout/transformers.rs:526]: When do we get this Canceled status? if this is during authentication then should be mapped to payment failure

File: crates/router/src/connector/checkout/transformers.rs [hunk 5]
```diff
@@ -514,10 +540,16 @@ impl ForeignFrom<(CheckoutPaymentStatus, CheckoutPaymentIntent)> for enums::Atte
...
```
Comment (ArjunKarthik) [inline crates/router/src/connector/checkout/transformers.rs:552]: When do we get this Canceled status? if this is during authentication then should be mapped to payment failure

File: crates/router/src/connector/checkout/transformers.rs [hunk 6]
```diff
@@ -537,10 +569,16 @@ impl ForeignFrom<(CheckoutPaymentStatus, Option<Balances>)> for enums::AttemptSt
...
```
Comment (ArjunKarthik) [inline crates/router/src/connector/checkout/transformers.rs:581]: When do we get this Canceled status? if this is during authentication then should be mapped to payment failure

File: crates/router/src/connector/checkout/transformers.rs [hunk 7]
```diff
@@ -1220,13 +1270,22 @@ pub enum CheckoutDisputeTransactionType {
...
+            | CheckoutWebhookEventType::AuthenticationFailed
+            | CheckoutWebhookEventType::PaymentAuthenticationFailed => {
+                Self::PaymentIntentAuthorizationFailure
+            }
             CheckoutWebhookEventType::PaymentApproved => Self::EventNotSupported,
             CheckoutWebhookEventType::PaymentCaptured => Self::PaymentIntentSuccess,
             CheckoutWebhookEventType::PaymentDeclined => Self::PaymentIntentFailure,
             CheckoutWebhookEventType::PaymentRefunded => Self::RefundSuccess,
             CheckoutWebhookEventType::PaymentRefundDeclined => Self::RefundFailure,
+            CheckoutWebhookEventType::PaymentCanceled => Self::PaymentIntentCancelled,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/checkout/transformers.rs:1286]: When do we get this Canceled status? if this is during authentication then should be mapped to payment failure

Comment (swangi-kumari) [reply crates/router/src/connector/checkout/transformers.rs:1286]: Occurs when the customer has canceled the payment via a provider's platform.

**Review by ArjunKarthik — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Pull request merged**